awx/main/tasks/callback.py

```diff
@@ -32,10 +34,26 @@ def __init__(self, model=None):
         self.safe_env = {}
         self.event_ct = 0
         self.model = model
+        self.extra_update_fields = {}
 
     def update_model(self, pk, _attempt=0, **updates):
         return update_model(self.model, pk, _attempt=0, **updates)
 
+    def delay_update(self, **kwargs):
```

Comment:
By caching artifact updates on the in-memory object like this, won't we be making more pain for ourselves in the future when we try to move to an async style of processing jobs?

Reply:
If we want to wait until the end of the job, is it possible to do a final sweep over the job events for that job and accumulate the data that way?  Or do we lose this data?

Reply:
We can save things on the job record and job events in one of two places
 - the control process
 - the callback receiver

The problem with the callback receiver is that it can (notoriously) be delayed as the events processing is the long-tail action, where a large queue can build up because we don't have the throughput to save at the rate we can produce events. The job _status_ can flip long before all of its events come in. This may have nothing to do with the output of that job. Events may come in at a delayed time before the _last job_ before that job produced more events than could be immediately handled, so all new events have to wait in line.

We cannot have artifacts wait in line with other events. The status changes, and dependent jobs in workflows will start with or without the events saved, and these can use artifacts. Likewise, the event_ct is needed to figure out if all events have been saved in the first place.